### ISSUE TITLE
Cannot read properties of undefined (reading 'cdnURL')

### ISSUE THREAD

**Issue by AlexanderBredun** [nuxt-modules/storybook#753]
I used this [Nuxt Storybook starter](https://github.com/nuxt-modules/storybook/tree/main/examples/starter), but when I try to add any image from the public directory, I get the following error:
![image](https://github.com/user-attachments/assets/50eda142-a3f7-40eb-a018-b82794e01292)
When I checked the console, it pointed me to this part of the code and showed that there is a problem with useRuntimeConfig:
![image](https://github.com/user-attachments/assets/5bfeedb0-97d2-4a23-9ce5-a95700a57dec)

Node 18.18.2, npm 9.8.1, windows 10

**Comment by guim4dev** [comment]
@AlexanderBredun did you manage to find a workaround?

**Comment by obulat** [comment]
To replicate, build the showcase app that has `/pages` in it:

`pnpm dev:prepare & pnpm example:showcase:storybook:build`

Go to the build folder (`storybook/examples/showcase/storybook-static`) and run an http server in it:
`npx http-server`

Check the logs in the console.

**Comment by tobiasdiez** [comment]
Thanks @obulat for the reproduction. This only occurs in the static build, and not when looking in console for the dev build ( `example:showcase:dev` > Storybook), right? 

That makes debugging this a bit harder...

**Comment by obulat** [comment]
> Thanks @obulat for the reproduction. I get the error there now as well. This only occurs in the static build, and not when looking in console for the dev build ( `example:showcase:dev` > Storybook), right?
> 
> That makes debugging this a bit harder...

I think that this is related to #761, and the errors there can be masking these errors. router is definitely part of the problem here.

**Comment by doppelmutzi** [comment]
I have a similar problems. Occurs with the autogenerated Docs view when I have multiple exports in a stories file.

**Comment by guim4dev** [comment]
adding some extra info here

@doppelmutzi I had the same problem. It only occurs in autogenerated docs with  >1 stories. Single Story autogenerated docs work fine, its probably something related to routing as pointed out by @obulat

I disabled autodocs for now as multi stories are important for our use case and it works fine locally.

Sadly now im getting blocked by chromatic, as the built version of storybook also throws the cdnURL error :/

**Comment by d-low** [comment]
Could this be a race condition? I set a breakpoint on line 2 in paths.mjs and then stepped into `useRuntimeConfig()`:

```JavaScript
// /.nuxt/nitro.client.mjs
export const useRuntimeConfig = () => 
window?.__NUXT__?.config || {}
```

In `useRuntimeConfig()` `window.__NUXT__` was undefined.

However, when I later check whether this global variable is defined in the console I can see that it is:

![Screenshot 2024-08-29 at 17 23 52](https://github.com/user-attachments/assets/6fde112f-e6c3-4ee7-8b67-345ff8aa6dd3)

**Comment by enkot** [comment]
@tobiasdiez Thank you for your work on this module 💚 

It would be great to prioritise this issue. In my case I can't upgrade to Nuxt 3.13.x because of this error.

I will also try to investigate.

**Comment by Limonische** [comment]
Encountered this error with latest versions of nuxt and storybook.
This helped resolve the issue:
```
vite: {
    resolve: {
      alias: {
        vue: 'vue/dist/vue.esm-bundler.js',
      },
    },
}
```

**Comment by tobiasdiez** [comment]
Thanks for looking into this. If it's indeed a problem with the runtime bundler of vue, then this should be fixed as part of https://github.com/nuxt-modules/storybook/pull/740 (which sadly is stuck in another issue).

**Comment by bojanrajh** [comment]
As a workaround, because we are not using cdn or changing the base URL, `pnpm patch nuxt` did the trick, for now:

```diff
diff --git a/dist/index.mjs b/dist/index.mjs
index 9fe137cf72916a0ceb3edbae12c9deacd960faed..625bb157338fbf10083760166600e518d81065c5 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -5119,7 +5119,8 @@ const publicPathTemplate = {
       "export const buildAssetsDir = () => appConfig.buildAssetsDir",
       "export const buildAssetsURL = (...path) => joinRelativeURL(publicAssetsURL(), buildAssetsDir(), ...path)",
       "export const publicAssetsURL = (...path) => {",
-      "  const publicBase = appConfig.cdnURL || appConfig.baseURL",
+      "  const publicBase = ''",
       "  return path.length ? joinRelativeURL(publicBase, ...path) : publicBase",
       "}",
       // On server these are registered directly in packages/nuxt/src/core/runtime/nitro/renderer.ts

```

**Comment by losbeekos** [comment]
When you use Storybook's render function to render out a component (or components) this issue arises as well.

**Comment by andresilva-cc** [comment]
@d-low for me `window.__NUXT__` is always undefined, even after some time.

I've been doing some tests and the function passed to the `setup` function inside `preview.ts` is never called, which explains why `window.__NUXT__` is always undefined. I see that the `setup` function is provided by `@storybook/vue3`, which, according to their code, adds the passed function into `globalThis.PLUGINS_SETUP_FUNCTIONS` and it's later executed in `runSetupFunctions`, but, again, those are never called for me.

I'm having trouble understanding how the 2 packages work together. I know that the Nuxt renderer is a wrapper for `@storybook/vue3`, but things start to get complicated inside `@storybook/vue3`, where it seems the code reaches only the `preset.js` where it provides the `previewAnnotations` const, but the provided files (`entry-preview.mjs` and `entry-preview-docs.mjs`) never seem to be called.

Does anyone have an idea of what's happening?

**Comment by andresilva-cc** [comment]
Reverting commit bcf68df stops the error, but images are broken, maybe it's just a matter of adding `staticDirs` to `.storybook/main.ts`.

Ok, so temporary workaround for anyone struggling with this. Add this snippet to your `.storybook/preview.ts`:

```ts
// @ts-expect-error internal Nuxt property
window.__NUXT__ = {
  config: {
    app: {
      baseURL: '/',
      cdnURL: ''
    }
  }
}
```

And to make images work, add this setting to your `.storybook/main.ts`:

```ts
staticDirs: ['../public'],
```

**Comment by JoJk0** [comment]
>  Reverting commit [bcf68df](https://github.com/nuxt-modules/storybook/commit/bcf68dff2ca3353733c4dc60f479f44b8c0cd15d) stops the error, but images are broken, maybe it's just a matter of adding `staticDirs` to `.storybook/main.ts`.
> 
> Ok, so temporary workaround for anyone struggling with this. Add this snippet to your `.storybook/preview.ts`:
> 
> // @ts-expect-error internal Nuxt property
> window.__NUXT__ = {
>   config: {
>     app: {
>       baseURL: '/',
>       cdnURL: ''
>     }
>   }
> }
> And to make images work, add this line to your config in `.storybook/main.ts`:
> 
> staticDirs: ['../public'],

Looks like this error is not showing up when I remove `router.options.ts` file so yeah definitely something connected to the router.
When I include routes and I apply `window.__NUXT__` patch from above, then it solves the issue, however, another one arises: 

![Image](https://github.com/user-attachments/assets/e018e299-28a8-4d01-93ff-2749343840fc)

**Comment by danieelbog** [comment]
Any new on this? I also have what JoJk0 says

**Comment by etsubeta132** [comment]
@AlexanderBredun  Is the issue fixed ?

**Comment by yanniznik** [comment]
Same issue, interested in solution!

**Comment by GonzaloGPF** [comment]
I have same issue, with just fresh installation.
Also, if I create a simple **app-button.stories.ts**, it throws errors too

![Image](https://github.com/user-attachments/assets/24bdc4f1-326f-498c-bbcb-9e49e9210190)

**Comment by pmaiko** [comment]
This issue occurs when running Storybook in development using the command storybook dev. To fix it, run IS_STORYBOOK=true storybook dev and add pages: !process.env.IS_STORYBOOK in the Nuxt config.

**Comment by bojanrajh** [comment]
This is our current `pnpm patch nuxt`.

```diff
diff --git a/dist/index.mjs b/dist/index.mjs
index 7fa92a5cd8f111907c6d991a367d1442eb5c920f..ed13c46841bbaeff30947d4960dc55eddb240b26 100644
--- a/dist/index.mjs
+++ b/dist/index.mjs
@@ -6452,13 +6452,13 @@ const publicPathTemplate = {
     return [
       "import { joinRelativeURL } from 'ufo'",
       !nuxt.options.dev && "import { useRuntimeConfig } from 'nitropack/runtime'",
-      nuxt.options.dev ? `const getAppConfig = () => (${JSON.stringify(nuxt.options.app)})` : "const getAppConfig = () => useRuntimeConfig().app",
-      "export const baseURL = () => getAppConfig().baseURL",
-      "export const buildAssetsDir = () => getAppConfig().buildAssetsDir",
+      nuxt.options.dev ? `const getAppConfig = () => (${JSON.stringify(nuxt.options.app)})` : "const getAppConfig = () => useRuntimeConfig()?.app",
+      "export const baseURL = () => getAppConfig()?.baseURL || '/'",
+      "export const buildAssetsDir = () => getAppConfig()?.buildAssetsDir",
       "export const buildAssetsURL = (...path) => joinRelativeURL(publicAssetsURL(), buildAssetsDir(), ...path)",
       "export const publicAssetsURL = (...path) => {",
       "  const appConfig = getAppConfig()",
-      "  const publicBase = appConfig.cdnURL || appConfig.baseURL",
+      "  const publicBase = appConfig?.cdnURL || appConfig?.baseURL || '/'",
       "  return path.length ? joinRelativeURL(publicBase, ...path) : publicBase",
       "}",
       // On server these are registered directly in packages/nuxt/src/core/runtime/nitro/handlers/renderer.ts
```

**Comment by AndrewBogdanovTSS** [comment]
@bojanrajh, the patch worked for me, but do you know if there is any issue inside Nuxt that would address it in the Core, so we won't need to support the patch internally?
